### ISSUE TITLE
Exporting fields with changed `column_name` doesn't work in v4

### ISSUE THREAD

**Issue by PetrDlouhy** [django-import-export/django-import-export#1846]
**Describe the bug**
Fields like `email = Field(column_name="recipientEmail")` from my `ModelResource` are missing in the CSV.

**To Reproduce**
1. Add `ModelResource` like:
```python
class WisePayoutResource(ModelResource):
    note_recipient = Field(column_name="name")
    email = Field(column_name="recipientEmail")
    id = Field(column_name="paymentRefernece")
    reciever_type = Field(column_name="recieverType")
    amount_currency = Field(column_name="amountCurrency")
    amount = Field(column_name="amount")
    source_currency = Field(column_name="sourceCurrency")
    target_currency = Field(column_name="targetCurrency")
    type = Field(column_name="type")

    class Meta:
        model = models.Payout
        export_order = (
            "note_recipient",
            "email",
            "id",
            "reciever_type",
            "amount_currency",
            "amount",
            "source_currency",
            "target_currency",
            "type",
        )

    def dehydrate_source_currency(self, payout):
        return "USD"

    def dehydrate_target_currency(self, payout):
        return "USD"

    def dehydrate_amount_currency(self, payout):
        return "source"

    def dehydrate_note_recipient(self, payout):
        return NOTE_RECIPIENT

    def dehydrate_reciever_type(self, payout):
        return "PERSON"

    def dehydrate_email(self, payout):
        return payout.get_payout_data()["payout_email"]

    def dehydrate_amount(self, payout):
        return payout.amount

    def dehydrate_id(self, payout):
        return payout.id

    def dehydrate_type(self, payout):
        return "EMAIL"

```

2. Export the table (e.g. Payout) with this resource and all columns enabled.
3. In the resulting CSV are only columns `amount` and `type` (those that didn't change name of the column).

**Versions (please complete the following information):**
 - Django Import Export: 4.0.3
 - Python 3.11
 - Django 4.2

**Expected behavior**
All columns should be exported.

**Additional context**
Seems that the column name is not correctly matched in this case. @matthewhegarty I expect this would be easy for you to find the problem and fix it, but if not I can do some more digging and fixing.

**Comment by matthewhegarty** [comment]
@PetrDlouhy - I can't reproduce it exactly in the way you have it.

- What happens if you declare a `fields` list with the fields you want in the export?
- I have a test case [here](https://github.com/django-import-export/django-import-export/blob/88eb0d3761b67a03c95515ce8ab46e11b7aebf66/tests/core/tests/test_resources/test_import_export.py#L256)
  - I find that if I remove the `fields` declaration, I get all Resource fields in the export. 
  - Are you able to reproduce using this test case?

**Comment by PetrDlouhy** [comment]
@matthewhegarty I think, that test is too specific for to test the actual problem. The test runs only `ModelResource.export()`, but the problem would probably lie in the admin form handling these field checkboxes:
![Snímek obrazovky_2024-05-27_16-02-11](https://github.com/django-import-export/django-import-export/assets/156755/83672457-cce8-488a-bfaa-ff98ec3ec190)


I will try to investigate further and make the appropriate test.

**Comment by matthewhegarty** [comment]
Hi Petr
I tried with your branch but reverted all your changes from `resources.py`.  I added a new EBook and set the author to "Ian Fleming".

When I export the Ebook, I see:

![image](https://github.com/django-import-export/django-import-export/assets/6249838/5e894f0b-f553-4c4a-87cf-2855e44f6fec)

The export file I get is:

```
id,Email of the author,name,published_date
1,test@example.com,Moonraker 2,2024-05-27
```

So it's possibly a bug that the Export HTML doesn't show the name of the field name to be export ("Email of the author"), but other than that it looks ok... Am I missing something?

**Comment by PetrDlouhy** [comment]
@matthewhegarty I didn't touch that part and I am not sure, which variant is correct. The original field name can be more readable for the user as it is/can be consistent across multiple resources.
I can imagine, that some resources would have totally unreadable column_names, e.g. IDs or legal identifiers.

Maybe we should display both values like `Author Email (Email of the author)`.

**Comment by matthewhegarty** [comment]
I notice that if I add the attribute name, then the test passes:

```
# test fails
author = fields.Field(column_name="AuthorFooName")

# test passes
author = fields.Field(attribute="author", column_name="AuthorFooName")
```

This is similar to what was reported [here](https://github.com/django-import-export/django-import-export/issues/1855#issuecomment-2130478806) this could be a v4 regression, I'll test in v3.

**Comment by matthewhegarty** [comment]
I tested in v3 (added test [here](https://github.com/django-import-export/django-import-export/blob/a98c0157498b264e3aefc10168861a3cc54fc305/tests/core/tests/test_resources.py#L2876)).  This passes in v3, so we have a regression in v4 when the attribute name is not supplied.

**Comment by PetrDlouhy** [comment]
@matthewhegarty I don't understand entirely. The test in will pass after applying #1856 (not #1857).

Is there something that needs to be fixed in either of those commits?
But maybe I am missing something, I will look at it tomorrow with fresh mind.

**Comment by matthewhegarty** [comment]
Hi, I am still looking into it.  I think there are a couple of things going on with other parts of the code which need addressing.  I'll write a more detailed response tomorrow.

**Comment by matthewhegarty** [comment]
hi Petr
I adapted your changes and made a new PR (#1861).  I found that if I [set an attribute on the field][1] during initialization, then this resolves the issue without needed to modify `resources.py`.  I have kept your other changes in.

The reason this works is because the `author` field in `test_export_fields_column_name` does not contain a defined attribute, therefore it is ignored during export.

This also fixes an issue I uncovered whilst testing (#1860).  I propose that I merge this PR instead of yours.  What do you think?

[1]: https://github.com/matthewhegarty/django-import-export/blob/ca6a4e31ba8151777e8ac4b6c35b84959f1713f5/import_export/declarative.py#L54

**Comment by PetrDlouhy** [comment]
@matthewhegarty That seems to be a better solution. I tested it and it seems to be working.

The PR is missing the refactoring made in this commit: https://github.com/django-import-export/django-import-export/pull/1861/commits/9d2cab3f164dfaa05b218724a24d0a542e646ad4
Those changes are unrelated to the PR itself, so it is clear why they are not in the PR (although the commit is and then the changes are reverted, so I suggest it might be better to rebase it out).
I think that the refactoring would be still beneficial for `django-import-export` and I can make it in separate PR, if you think it too.

**Comment by matthewhegarty** [comment]
I only removed the change so that I could be sure that changes to `resource.py` did not affect the fix in #1861.

There is some duplication in those methods, so feel free to submit a new PR to make the code clearer.  I would recommend to hide any shared logic in a protected method.

I also wasn't comfortable with the new Field attribute called `original_name`, but we shouldn't need that now due to the change in #1861.

Thanks again for all your help in identifying and fixing this issue :+1:

**Comment by PetrDlouhy** [comment]
Yeah, the `original_name` was only a tool to transfer the field variable name. Since now it is in `attribute` we don't need it anymore. I will make the refactor PR after #1861 is merged.

**Comment by PetrDlouhy** [comment]
@matthewhegarty Everything seems to be working. Thank you very much for your attention and all the work you put into this project.

**Comment by matthewhegarty** [comment]
np - thanks for your input!
